src/ops/base.rs

```diff
@@ -1,39 +1,221 @@
+use serde_json;
+use snowflake::ProcessUniqueId;
+use buf_redux::BufWriter;
+use buf_redux::strategy::WhenFull;
 use std::collections::HashMap;
+use std::fs;
+use std::fs::{File, OpenOptions};
+use std::path::PathBuf;
+use std::time::{Duration, Instant};
+use time;
+
+// 4k buffered log.
+const LOG_BUFFER_CAPACITY: usize = 4 * 1024;
+
+// We store at most this many write records before flushing to disk.
+const BUFFERED_WRITES_CAPACITY: usize = 512;
+
+// We spend at least this many milliseconds without flushing write records to disk.
+const BUFFERED_WRITES_FLUSH_INTERVAL_MS: u64 = 1000;
 
 /// Base is used to represent the root nodes of the distributary data flow graph.
 ///
 /// These nodes perform no computation, and their job is merely to persist all received updates and
 /// forward them to interested downstream operators. A base node should only be sent updates of the
 /// type corresponding to the node's type.
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct Base {
+    buffered_writes: Option<Records>,
+    durability: BaseDurabilityLevel,
+    durable_log: Option<BufWriter<File, WhenFull>>,
+    durable_log_path: Option<PathBuf>,
+    global_address: Option<NodeAddress>,
+    last_flushed_at: Option<Instant>,
```

Comment:
Oh, so this stems from the same issue below of not knowing the node's ID until later. That's unfortunate. I think I'd still prefer `Option`s inside of a `BaseDurabilityState`, but you're right that it's less elegant. I wonder if there's a way we could name `Base` nodes when they're created instead of relying on `self.us`. That would mean we could instantiate all of this at creation time, get rid of the `Option`s, and also get rid of the call to `ensure_log`...